proxy/api/client.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"net"
+)
+
+type Client struct {
+	conn    *net.UnixConn
+	encoder *json.Encoder
+	decoder *json.Decoder
+}
+
+// Creates a new client object to communicate with the proxy using the
+// connection conn
+func NewClient(conn *net.UnixConn) *Client {
+	return &Client{
+		conn:    conn,
+		encoder: json.NewEncoder(conn),
+		decoder: json.NewDecoder(conn),
+	}
+}
+
+func (client *Client) Close() {
+	client.conn.Close()
+}
+
+func (client *Client) sendPayload(id string, payload interface{}) (*Response, error) {
+	var err error
+
+	req := Request{}
+	req.Id = id
+	if payload != nil {
+		if req.Data, err = json.Marshal(payload); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := client.encoder.Encode(&req); err != nil {
+		return nil, err
+	}
+
+	resp := Response{}
+	if err := client.decoder.Decode(&resp); err != nil {
+		return nil, err
+	}
+
+	return &resp, nil
+}
+
+func errorFromResponse(resp *Response) error {
+	// We should always have an error with the response, but better safe
+	// than sorry.
+	if resp.Success == false {
+		if resp.Error != "" {
+			return errors.New(resp.Error)
+		}
+
+		return errors.New("unknown error")
+	}
+
+	return nil
+}
+
+func (client *Client) Hello(containerId, ctlSerial, ioSerial string) error {
+	hello := Hello{
+		ContainerId: containerId,
+		CtlSerial:   ctlSerial,
+		IoSerial:    ioSerial,
+	}
+
```

Comment:
Right I thought about that. But then allocateIo is needed for two things:

  - start/create (hello + allocateIo)
  - exec (attach + allocateIo)

I've separated the initial workload that does an "hello" from exec workload that only do an "attach". Because of that, I added allocateIo as an orthogonal command that can be done in either case.

We could still decide to compress the messages in one (hello and attach could both ask to allocate the streams), may have some boot time gains.